Doppler.ContactPolicies.Api.Test/UpdateContactPoliciesSettingsTest.cs

```diff
@@ -0,0 +1,164 @@
+            };
+
+            // Act
+            var response = await client.SendAsync(request);
+
+            // Assert
+            Assert.NotNull(response);
+            Assert.Equal(HttpStatusCode.OK, response.StatusCode);
+        }
+
+        [Fact]
+        public async Task
+            UpdateContactPoliciesSettings_Should_ReturnInternalServerError_When_UserWithSameAccountNameIsFoundButDoesNotHavePermissions()
+        {
+            // Arrange
+            const string contactPoliciesRequestBodyStub = CONTACT_POLICIES_SETTINGS_REQUEST_BODY_STUB;
+            const string validAccountName = "test1@test.com";
+            const string token = TOKEN_ACCOUNT_123_TEST1_AT_TEST_DOT_COM_EXPIRE_20330518;
+
+            var fixture = new Fixture();
+            var foundedIdUser = fixture.Create<int>();
+            var contactPoliciesSettings = new ContactPoliciesSettingsDto
+            {
+                AccountName = validAccountName,
+                IntervalInDays = 10,
+                Active = true,
+                EmailsAmountByInterval = 5,
+                ExcludedSubscribersLists = new List<ExcludedSubscribersLists>()
+            };
+            var contactPoliciesSettingDao = contactPoliciesSettings.ToDao();
+
+            // to allow throw exceptions
+            var contactPoliciesRepositoryMock = new Mock<IContactPoliciesSettingsRepository>(MockBehavior.Strict);
+
+            contactPoliciesRepositoryMock.Setup(x => x.GetIdUserByAccountName(validAccountName))
+                .ReturnsAsync(foundedIdUser);
+            contactPoliciesRepositoryMock.Setup(x => x.UpdateContactPoliciesSettingsAsync(foundedIdUser, contactPoliciesSettingDao))
+                .ThrowsAsync(new Exception("This action is not allowed for this user."));
+
+            var contactService = new ContactPoliciesService(contactPoliciesRepositoryMock.Object);
+
+            var client = _factory.WithWebHostBuilder((e) => e.ConfigureTestServices(services =>
+            {
+                services.AddSingleton(contactPoliciesRepositoryMock.Object);
+                services.AddSingleton(contactService);
+            })).CreateClient();
+            var request = new HttpRequestMessage(HttpMethod.Put, $"/accounts/{validAccountName}/settings")
+            {
+                Headers = { { "Authorization", $"Bearer {token}" } },
+                Content = new StringContent(contactPoliciesRequestBodyStub, Encoding.UTF8, "application/json")
+            };
+
+            // Act
+            var response = await client.SendAsync(request);
+
+            // Assert
+            Assert.NotNull(response);
+            var expectedRepositoryException = await Assert.ThrowsAsync<Exception>(() =>
+                contactPoliciesRepositoryMock.Object.UpdateContactPoliciesSettingsAsync(foundedIdUser, contactPoliciesSettingDao));
+            Assert.Contains("This action is not allowed for this user.", expectedRepositoryException.Message);
```

Comment:
Tiene sentido comprobar un mensaje que se está agregando al mock?

Reply:
La idea aquí sería comprobar el comportamiento para el caso en que exista un usuario pero este no tenga permisos para hacer cambios en la politica de contactos. Estaríamos comprobando que si este escenario sucede entonces debemos lanzar una excepción con el mensaje "This action is not allowed for this user.". No encontré otra forma de mockear la excepción que sale del repositorio. Y la forma que encontré de comprobar la excepción es la que implementé, en los Asserts quedarme con la excepción que va a salir del repositorio mockeado y validar el mensaje de la misma

Reply:
@jkortega9008 @andresmoschini 
Qué opinan de reducir el alcance del test a que se debe retornar un `InternalServerError` ante excepción en el repositorio ? (No testear de esta manera la parte de permisos)

Reply:
No entiendo lo que me estas proponiendo con este cambio 😞

Reply:
> Qué opinan de reducir el alcance del test a que se debe retornar un InternalServerError ante excepción en el repositorio ? (No testear de esta manera la parte de permisos)

Me parece que está bien, porque lo que hay ahora es una tautologia: mockeo algo y testeo lo que acabo de mockear.

> No entiendo lo que me estas proponiendo con este cambio 😞

Creo que no hay forma de testear el escenario real, por eso Darian propone testear solo lo testeable.

Reply:
ok podriamos dejar solo estas líneas de código en los Assert:
`// Assert
            Assert.NotNull(response);
            var expectedRepositoryException = await Assert.ThrowsAsync<Exception>(() =>
                contactPoliciesRepositoryMock.Object.UpdateContactPoliciesSettingsAsync(foundedIdUser, contactPoliciesSettingDao));
            Assert.Contains("This action is not allowed for this user.", expectedRepositoryException.Message);
            Assert.Equal(HttpStatusCode.InternalServerError, response.StatusCode);`
y eliminamos el mock que se le hace al repositorio devolviendo la excepción

Reply:
Hablamos con @jkortega9008, va a reducir el alcance del test a comprobar que el request retorne `InternalServerError` ante cualquier excepción en el repositorio.